[inductor] Allow reinplacing functionalized scatter ops

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116899
* #117121
* #116898

This expands the reinplacing pass to allow reinplacing view-scatter operations.
e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```

First, the `canonicalize_view_scatter_ops` step rewrites the functionalized graph 
in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism. Since we've fused the two scatter ops into one,
the reinplacing pass sees only one user of `inp` which allows the entire operation to be 
reinplaced  if desired (and I add heuristics that sometimes choose not to reinplace).

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations either back into view_scatter operations, or
into the version with mutations. When introducing mutations, the reinplaced
version is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place we end up with a minor restructuring of the graph:
```
a = view1(inp)
tmp = view2_scatter(a, src)
inp_updated = view1_scatter(inp, tmp)
a_updated = view1(inp_updated)
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

peterbell10 committed:
[inductor] Improved handling of functionalized scatter ops

This improves the reinplacing pass so we can now reinplace consecutive
view-scatter operations. e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```
previously we could only reinplace the last scatter since the first comes before
another use of `inp`.

To get around this, I add the `canonicalize_view_scatter_ops` step which rewrites
the functionalized graph in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism, but since there is only one use of `inp`
it is now successfully reinplaced.

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations into the version with mutations which when
reinplaced is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place it still uses fewer copy operations than the original,
```
inp_updated = clone(inp)
a = view1(inp_updated)
b = view2(a)
b.copy_(src)
```

[ghstack-poisoned]

leslie-fang-intel commented:
Here is a example https://gist.github.com/leslie-fang-intel/85a7876cd4097479b6507d20a0696874 which comes from llama https://github.com/pytorch/benchmark/blob/79c236aed69907988b941e730965e6bfc9fd8c21/torchbenchmark/models/llama/model.py#L124-L133

- FX graph transformed as baseline: https://gist.github.com/leslie-fang-intel/c1715eea80038d0c71f0763cfb865cc1
- FX graph transformed after https://github.com/pytorch/pytorch/pull/116789 is: https://gist.github.com/leslie-fang-intel/9bb5ef812ae19e89e0de3f177bf4947c
```
        slice_1: "f32[1, 384, 8, 64]" = torch.ops.aten.slice.Tensor(arg1_1, 0, 0, 1)
        slice_2: "f32[1, 32, 8, 64]" = torch.ops.aten.slice.Tensor(slice_1, 1, 1, 33)
        copy: "f32[1, 32, 8, 64]" = torch.ops.aten.copy.default(slice_2, view_1);  slice_2 = view_1 = None
        slice_scatter: "f32[1, 384, 8, 64]" = torch.ops.aten.slice_scatter.default(slice_1, copy, 1, 1, 33);  slice_1 = copy = None
        
        # No stacktrace found for following nodes
        slice_tensor: "f32[1, 384, 8, 64]" = torch.ops.aten.slice.Tensor(arg1_1, 0, 0, 1, 1)
        copy__default: "f32[1, 384, 8, 64]" = torch.ops.aten.copy_.default(slice_tensor, slice_scatter);  slice_tensor = slice_scatter = None
```
- FX graph transformed with current PR is: https://gist.github.com/leslie-fang-intel/87a6fee627cca048e57677a9f4e381ff
```
        # File: /home/jianan/leslie/torch_inductor_lz/torch_script/inductor/fp32/kv_cache_code_gen_issue/test_kv_cache_code_gen.py:24, code: self.cache_k[:bsz, start_pos : start_pos + seqlen] = xk
        slice_1: "f32[1, 384, 8, 64]" = torch.ops.aten.slice.Tensor(arg1_1, 0, 0, 1)
        slice_2: "f32[1, 32, 8, 64]" = torch.ops.aten.slice.Tensor(slice_1, 1, 1, 33);  slice_1 = None
        copy: "f32[1, 32, 8, 64]" = torch.ops.aten.copy.default(slice_2, view_1);  slice_2 = view_1 = None
        
        # No stacktrace found for following nodes
        clone: "f32[56, 384, 8, 64]" = torch.ops.aten.clone(arg1_1)
        slice_tensor: "f32[1, 384, 8, 64]" = torch.ops.aten.slice.Tensor(clone, 0, 0, 1)
        slice_tensor_1: "f32[1, 32, 8, 64]" = torch.ops.aten.slice.Tensor(slice_tensor, 1, 1, 33);  slice_tensor = None
        copy__1: "f32[1, 32, 8, 64]" = torch.ops.aten.copy_(slice_tensor_1, copy);  slice_tensor_1 = copy = None
----------------------
        copy_: "f32[56, 384, 8, 64]" = torch.ops.aten.copy_.default(arg1_1, clone);  arg1_1 = clone = None
```

I think it's not as efficient as https://github.com/pytorch/pytorch/pull/116789, may because we failed to reinplace the `slice_scatter` node in this case.

peterbell10 committed:
Fix _generalized_scatter fake impl on "[inductor] Improved handling of functionalized scatter ops"


This improves the reinplacing pass so we can now reinplace consecutive
view-scatter operations. e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```
previously we could only reinplace the last scatter since the first comes before
another use of `inp`.

To get around this, I add the `canonicalize_view_scatter_ops` step which rewrites
the functionalized graph in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism, but since there is only one use of `inp`
it is now successfully reinplaced.

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations into the version with mutations which when
reinplaced is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place it still uses fewer copy operations than the original,
```
inp_updated = clone(inp)
a = view1(inp_updated)
b = view2(a)
b.copy_(src)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented:
There was a small bug in the fake tensor propagation of `_generalized_scatter` which caused its result to be treated as an alias of the input, which was preventing reinplacing. With that fixed the cache example is successfully reinplaced.

peterbell10 committed:
Only reinplace scatters when profitable on "[inductor] Improved handling of functionalized scatter ops"


This improves the reinplacing pass so we can now reinplace consecutive
view-scatter operations. e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```
previously we could only reinplace the last scatter since the first comes before
another use of `inp`.

To get around this, I add the `canonicalize_view_scatter_ops` step which rewrites
the functionalized graph in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism, but since there is only one use of `inp`
it is now successfully reinplaced.

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations into the version with mutations which when
reinplaced is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place it still uses fewer copy operations than the original,
```
inp_updated = clone(inp)
a = view1(inp_updated)
b = view2(a)
b.copy_(src)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Allow reinplacing functionalized scatter ops"


This expands the reinplacing pass to allow reinplacing view-scatter operations.
e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```

First, the `canonicalize_view_scatter_ops` step rewrites the functionalized graph 
in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism. Since we've fused the two scatter ops into one,
the reinplacing pass sees only one user of `inp` which allows the entire operation to be 
reinplaced  if desired (and I add heuristics that sometimes choose not to reinplace).

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations either back into view_scatter operations, or
into the version with mutations. When introducing mutations, the reinplaced
version is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place we end up with a minor restructuring of the graph:
```
a = view1(inp)
tmp = view2_scatter(a, src)
inp_updated = view1_scatter(inp, tmp)
a_updated = view1(inp_updated)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Allow reinplacing functionalized scatter ops"


This expands the reinplacing pass to allow reinplacing view-scatter operations.
e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```

First, the `canonicalize_view_scatter_ops` step rewrites the functionalized graph 
in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism. Since we've fused the two scatter ops into one,
the reinplacing pass sees only one user of `inp` which allows the entire operation to be 
reinplaced  if desired (and I add heuristics that sometimes choose not to reinplace).

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations either back into view_scatter operations, or
into the version with mutations. When introducing mutations, the reinplaced
version is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place we end up with a minor restructuring of the graph:
```
a = view1(inp)
tmp = view2_scatter(a, src)
inp_updated = view1_scatter(inp, tmp)
a_updated = view1(inp_updated)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Allow reinplacing functionalized scatter ops"


This expands the reinplacing pass to allow reinplacing view-scatter operations.
e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```

First, the `canonicalize_view_scatter_ops` step rewrites the functionalized graph 
in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism. Since we've fused the two scatter ops into one,
the reinplacing pass sees only one user of `inp` which allows the entire operation to be 
reinplaced  if desired (and I add heuristics that sometimes choose not to reinplace).

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations either back into view_scatter operations, or
into the version with mutations. When introducing mutations, the reinplaced
version is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place we end up with a minor restructuring of the graph:
```
a = view1(inp)
tmp = view2_scatter(a, src)
inp_updated = view1_scatter(inp, tmp)
a_updated = view1(inp_updated)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented:
[Benchmark results are in](https://hud.pytorch.org/benchmark/compilers?startTime=Mon%2C%2008%20Jan%202024%2001%3A37%3A16%20GMT&stopTime=Mon%2C%2015%20Jan%202024%2001%3A37%3A16%20GMT&granularity=hour&suite=torchbench&mode=inference&dtype=bfloat16&lBranch=gh/peterbell10/679/orig&lCommit=29d4fcfd2f7f8ad87c13cbc7e0f6aa983a37d391&rBranch=main&rCommit=d78776e2e60ae0257847097dda3eb2fdc6f32476). This only seems to effect llama inference, but that speedup is very significant. default went from 1.442x -> 1.600x speedup, cudagraphs from 5.539x -> 8.994x, and cudagraphs_freezing from 6.319x -> 11.146x.

There are a few other changes reported by the benchmark suite but I haven't been able to replicate and in every case I checked the fx graphs didn't even contain any `_scatter` ops.

lezcano commented on torch/_inductor/fx_passes/reinplace.py:
```diff
@@ -1,24 +1,331 @@
+            kwargs=node.kwargs,
+        )
+
+        if can_fuse := src.target is _generalized_scatter:
```
This logic is a bit convoluted (redefining `can_fuse` in the if). Can you factor it out into a helper function?

lezcano commented on torch/_inductor/fx_passes/reinplace.py:
```diff
@@ -1,24 +1,331 @@
+                scatter_view_op,
+            ]
+
+        if not can_fuse:
```
I don't quite understand this part. What happens here if you simply have a view -> view -> scatter? In this case we cannot fuse, but we are creating a generalized_scatter with just one view rather than two? Shouldn't we also add `*node_to_view_op[src]`?

peterbell10 replied:
Lets say for example we have:
```python
def fn(a, src):
    a_view = a.as_strided(new_size, new_stride)
    a_view_updated = aten.slice_scatter(a_view, src, 1, 10, None)
    # ... use a_view_updated
```
IIUC your suggestion is to canonicalize this as:
```python
def fn(a, src):
    a_updated = _generalized_scatter(a, src, [
        ViewOp(aten.as_strided, (new_size, new_stride), {}), 
        ViewOp(aten.slice, (1, 10, None), {})
    ])
    a_view_updated = a_updated.as_strided(new_size, new_stride)
    # ...
```
where the scatter op is now in terms of the base tensor instead of the original scatter op argument. 

In this (cherry-picked) example it's rather problematic because `as_strided` forces the buffer to be realized and the scatter to be handled via mutation, whereas the original `slice_scatter` doesn't require mutation. Also even if the `as_strided` only views a subset of `a`, this version would still need to copy the entire buffer.

As written, this PR instead canonicalizes the example as:
```python
def fn(a, src):
    a_view = a.as_strided(new_size, new_stride)
    a_view_updated = _generalized_scatter(a_view, src, [ViewOp(aten.slice, (1, 10, None), {})])
    # ...
```

Which we can directly translate back to the original `slice_scatter` and so no buffer gets realized.

peterbell10 committed:
Address review comments on "[inductor] Allow reinplacing functionalized scatter ops"


This expands the reinplacing pass to allow reinplacing view-scatter operations.
e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```

First, the `canonicalize_view_scatter_ops` step rewrites the functionalized graph 
in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism. Since we've fused the two scatter ops into one,
the reinplacing pass sees only one user of `inp` which allows the entire operation to be 
reinplaced  if desired (and I add heuristics that sometimes choose not to reinplace).

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations either back into view_scatter operations, or
into the version with mutations. When introducing mutations, the reinplaced
version is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place we end up with a minor restructuring of the graph:
```
a = view1(inp)
tmp = view2_scatter(a, src)
inp_updated = view1_scatter(inp, tmp)
a_updated = view1(inp_updated)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/fx_passes/reinplace.py:
```diff
@@ -1,24 +1,331 @@
+    return any(view.target in _ALWAYS_MUTATING_SCATTER_OPS for view in view_ops)
+
+
+def should_reinplace_scatter(node: torch.fx.Node) -> bool:
```
I've factored out the generic part and called it `is_node_realized`, though some of this function does assume that `node` is a `_generalized_scatter` so this function still has to exist.

peterbell10 committed:
Rebase on "[inductor] Allow reinplacing functionalized scatter ops"


This expands the reinplacing pass to allow reinplacing view-scatter operations.
e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```

First, the `canonicalize_view_scatter_ops` step rewrites the functionalized graph 
in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism. Since we've fused the two scatter ops into one,
the reinplacing pass sees only one user of `inp` which allows the entire operation to be 
reinplaced  if desired (and I add heuristics that sometimes choose not to reinplace).

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations either back into view_scatter operations, or
into the version with mutations. When introducing mutations, the reinplaced
version is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place we end up with a minor restructuring of the graph:
```
a = view1(inp)
tmp = view2_scatter(a, src)
inp_updated = view1_scatter(inp, tmp)
a_updated = view1(inp_updated)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Allow reinplacing functionalized scatter ops"


This expands the reinplacing pass to allow reinplacing view-scatter operations.
e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```

First, the `canonicalize_view_scatter_ops` step rewrites the functionalized graph 
in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism. Since we've fused the two scatter ops into one,
the reinplacing pass sees only one user of `inp` which allows the entire operation to be 
reinplaced  if desired (and I add heuristics that sometimes choose not to reinplace).

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations either back into view_scatter operations, or
into the version with mutations. When introducing mutations, the reinplaced
version is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place we end up with a minor restructuring of the graph:
```
a = view1(inp)
tmp = view2_scatter(a, src)
inp_updated = view1_scatter(inp, tmp)
a_updated = view1(inp_updated)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -r

peterbell10 committed:
Update on "[inductor] Allow reinplacing functionalized scatter ops"


This expands the reinplacing pass to allow reinplacing view-scatter operations.
e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```

First, the `canonicalize_view_scatter_ops` step rewrites the functionalized graph 
in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism. Since we've fused the two scatter ops into one,
the reinplacing pass sees only one user of `inp` which allows the entire operation to be 
reinplaced  if desired (and I add heuristics that sometimes choose not to reinplace).

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations either back into view_scatter operations, or
into the version with mutations. When introducing mutations, the reinplaced
version is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place we end up with a minor restructuring of the graph:
```
a = view1(inp)
tmp = view2_scatter(a, src)
inp_updated = view1_scatter(inp, tmp)
a_updated = view1(inp_updated)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 committed:
Lint on "[inductor] Allow reinplacing functionalized scatter ops"


This expands the reinplacing pass to allow reinplacing view-scatter operations.
e.g. if our python code is:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```
this generates a functionalized graph like:
```python
a = view1(inp)
a_updated = view2_scatter(a, src)
inp_updated = view1_scatter(inp, a_updated)
```

First, the `canonicalize_view_scatter_ops` step rewrites the functionalized graph 
in the form:
```python
inp_updated = _generalized_scatter(inp, src, [view1, view2])
a_updated = view1(inp_updated)
```

I then register `_generalized_scatter` as a normal inplacable op which can be
handled by the pre-existing mechanism. Since we've fused the two scatter ops into one,
the reinplacing pass sees only one user of `inp` which allows the entire operation to be 
reinplaced  if desired (and I add heuristics that sometimes choose not to reinplace).

Finally, there is a decomposition step which decomposes out-of-place or in-place
`_generalized_scatter` operations either back into view_scatter operations, or
into the version with mutations. When introducing mutations, the reinplaced
version is equivalent to the original mutation:
```
a = view1(inp)
b = view2(a)
b.copy_(src)
```

Or when out-of-place we end up with a minor restructuring of the graph:
```
a = view1(inp)
tmp = view2_scatter(a, src)
inp_updated = view1_scatter(inp, tmp)
a_updated = view1(inp_updated)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

JacobSzwejbka commented:
@peterbell10 Sorry to ping you so long after you merged this :P. Do you remember the difference between this pass and the one that isn't in inductor https://github.com/pytorch/pytorch/blob/main/torch/fx/passes/reinplace.py?
